src/System.Private.CoreLib/src/System/Array.cs

```diff
@@ -271,9 +271,7 @@ private static unsafe void CopyImpl(Array sourceArray, int sourceIndex, Array de
             if (destinationArray == null)
                 throw new ArgumentNullException("destinationArray");
 
-            int sourceRank = sourceArray.Rank;
-            int destinationRank = destinationArray.Rank;
-            if (sourceRank != destinationRank)
+            if (sourceArray.EETypePtr != destinationArray.EETypePtr && sourceArray.Rank != destinationArray.Rank)
```

Comment:
@jkotas What about a raw pointer comparison, `src.EETypePtr.ToPointer() == dst.EETypePtr.ToPointer()`? Is it going to be a frequent case that the memory location pointed to by both pointers of equal types will be the same?